core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1243,19 +1244,30 @@ class KafkaApis(val requestChannel: RequestChannel,
       topicResponses
     } else {
       val nonExistentTopics = topics.diff(topicResponses.map(_.name).toSet)
-      val responsesForNonExistentTopics = nonExistentTopics.map { topic =>
+      val responsesForNonExistentTopics = nonExistentTopics.flatMap { topic =>
         if (isInternal(topic)) {
           val topicMetadata = createInternalTopic(topic)
-          if (topicMetadata.errorCode == Errors.COORDINATOR_NOT_AVAILABLE.code)
-            metadataResponseTopic(Errors.INVALID_REPLICATION_FACTOR, topic, true, util.Collections.emptyList())
-          else
-            topicMetadata
+          Some(
+            if (topicMetadata.errorCode == Errors.COORDINATOR_NOT_AVAILABLE.code)
+              metadataResponseTopic(Errors.INVALID_REPLICATION_FACTOR, topic, true, util.Collections.emptyList())
+            else
+              topicMetadata
+          )
+        } else if (isFetchAllMetadata) {
+          // KAFKA-10606: If this request is to get metadata for all topics, auto topic creation should not be allowed
+          // The special handling is necessary on broker side because allowAutoTopicCreation is hard coded to true
+          // for backward compatibility on client side.
+          //
+          // However, in previous versions, UNKNOWN_TOPIC_OR_PARTITION won't happen on fetch all metadata,
+          // so, for backward-compatibility, we need to skip these not founds during fetch all metadata here.
```

Comment:
We don't usually include JIRA references in the code unless it's a very complex issue. I think the comment here could be something like:

"A metadata request for all topics should never result in topic auto creation. A topic may be deleted between the creation of `topics` and `topicResponses`, so we make sure to always return `None` for this case."

Reply:
Thanks @ijuma , 

To follow the convention in Kafka, I just updated the comments.  Let me know if you have there are any other thoughts!